engines/e_pkcs11.c

```diff
@@ -0,0 +1,504 @@
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_Initialize failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_INITIALIZE, PKCS11_R_INITIALIZE_FAILED);
+        return rv;
+    }
+
+    return CKR_OK;
+}
+
+static void pkcs11_finalize(void)
+{
+    pkcs11_funcs->C_Finalize(NULL);
+}
+
+int pkcs11_get_slot(PKCS11_CTX *ctx)
+{
+    CK_RV rv;
+    CK_ULONG slotCount;
+    CK_SLOT_ID slotId;
+    CK_SLOT_ID_PTR slotList;
+    unsigned int i;
+
+    rv = pkcs11_funcs->C_GetSlotList(CK_TRUE, NULL, &slotCount);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetSlotList failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, PKCS11_R_GET_SLOTLIST_FAILED);
+        goto err;
+    }
+
+    if (slotCount == 0) {
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, PKCS11_R_SLOT_NOT_FOUND);
+        goto err;
+    }
+
+    slotList = OPENSSL_malloc(sizeof(CK_SLOT_ID) * slotCount);
+
+    if (slotList == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_GetSlotList(CK_TRUE, slotList, &slotCount);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetSlotList failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_SLOT, PKCS11_R_GET_SLOTLIST_FAILED);
+        OPENSSL_free(slotList);
+        goto err;
+    }
+
+    slotId = slotList[0]; /* Default value if slot not set*/
+    for (i = 1; i < slotCount; i++) {
+        if (ctx->slotid == slotList[i]) slotId = slotList[i];
+    }
+
+    ctx->slotid = slotId;
+    OPENSSL_free(slotList);
+    return 1;
+
+ err:
+    return 0;
+}
+
+int pkcs11_start_session(PKCS11_CTX *ctx)
+{
+    CK_RV rv;
+    CK_SESSION_HANDLE session;
+
+    rv = pkcs11_funcs->C_OpenSession(ctx->slotid, CKF_SERIAL_SESSION, NULL,
+                                     NULL, &session);
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_OpenSession failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_START_SESSION,
+                  PKCS11_R_OPEN_SESSION_ERROR);
+        return 0;
+    }
+    ctx->session = session;
+    return 1;
+}
+
+int pkcs11_login(PKCS11_CTX *ctx, CK_USER_TYPE userType)
+{
+    /* Binary pins not supported */
+    CK_RV rv;
+
+    if (ctx->pin != NULL) {
+        rv = pkcs11_funcs->C_Login(ctx->session, userType, ctx->pin,
+                                   (CK_ULONG)strlen((char *)ctx->pin));
+        if (rv != CKR_OK) {
+            PKCS11_trace("C_Login failed, error: %#08X\n", rv);
+            PKCS11err(PKCS11_F_PKCS11_LOGIN, PKCS11_R_LOGIN_FAILED);
+            goto err;
+        }
+    return 1;
+    }
+ err:
+    return 0;
+}
+
+static int pkcs11_logout(CK_SESSION_HANDLE session)
+{
+    CK_RV rv;
+
+    rv = pkcs11_funcs->C_Logout(session);
+    if (rv != CKR_USER_NOT_LOGGED_IN && rv != CKR_OK) {
+        PKCS11_trace("C_Logout failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_LOGOUT, PKCS11_R_LOGOUT_FAILED);
+        goto err;
```

Comment:
Using a `goto err `in in this sort function don't improve its readability. beter just `return 0 `